### PR TITLE
Add per-pair collision safety margins

### PR DESCRIPTION
**Pull request by mpowelson**
Will lay the groundwork for #427

### CODE REVIEW

**Comment by mpowelson** [comment]
@Levi-Armstrong Some of the tests are failing. This seemed too easy. How does this look?

File: tesseract/tesseract_collision/include/tesseract_collision/core/common.h [hunk 1]
```diff
@@ -116,8 +124,11 @@ inline ContactResult* processResult(ContactTestData& cdata,
     ContactResultVector data;
     if (cdata.req.type == ContactTestType::FIRST)
     {
-      data.emplace_back(contact);
-      cdata.done = true;
+      if (contact.distance > cdata.safety_margin_data.getPairSafetyMarginData(key.first, key.second))
```
Comment (mpowelson) [inline tesseract/tesseract_collision/include/tesseract_collision/core/common.h:127]: Is this correct? I'm not sure if these keys align.

Comment (Levi-Armstrong) [reply tesseract/tesseract_collision/include/tesseract_collision/core/common.h:127]: Where is the definition of SafetyMarginData?

Comment (Levi-Armstrong) [reply tesseract/tesseract_collision/include/tesseract_collision/core/common.h:127]: It should be but I think we want to update the lookup table to work how the allowed collision matrix works.

Comment (mpowelson) [reply tesseract/tesseract_collision/include/tesseract_collision/core/common.h:127]: It is defined in types.h

Comment (mpowelson) [reply tesseract/tesseract_collision/include/tesseract_collision/core/common.h:127]: done

**Comment by mpowelson** [comment]
@Levi-Armstrong Well the tests are passing now.

**Comment by mpowelson** [comment]
It looks like somehow the contact data is not getting into the FCL distance callback. Bullet seems to be working so far.

**Comment by mpowelson** [comment]
@Levi-Armstrong I found the issue in FCL. I am cautiously optimistic that this is done when you get a chance to review it.

**Comment by mpowelson** [comment]
New profile data
```
Discrete: One Primitive Shape per link
Total number of shape: 1000
       BtSimple | DT:         0.06863 ms | Contacts: 50
          BtBVH | DT:         0.71992 ms | Contacts: 50
Discrete: All Primitive Shape in single link
Total number of shape: 1000
       BtSimple | DT:         0.27167 ms | Contacts: 50
          BtBVH | DT:         0.16516 ms | Contacts: 50
Discrete: One Convex Shape per link
Total number of shape: 1000
       BtSimple | DT:         1.06500 ms | Contacts: 50
          BtBVH | DT:         1.58933 ms | Contacts: 50
Discrete: All Convex Shape in single link
Total number of shape: 1000
       BtSimple | DT:         1.23065 ms | Contacts: 50
          BtBVH | DT:         1.07278 ms | Contacts: 50
Continuous: One Primitive Shape per link
Total number of shape: 1000
   BtCastSimple | DT:         2.78232 ms | Contacts: 50
      BtCastBVH | DT:         3.49627 ms | Contacts: 50
Continuous: All Primitive Shape in single link
Total number of shape: 1000
   BtCastSimple | DT:         3.26614 ms | Contacts: 50
      BtCastBVH | DT:         3.13047 ms | Contacts: 50
Continuous: One Convex Shape per link
Total number of shape: 1000
   BtCastSimple | DT:         1.21461 ms | Contacts: 50
      BtCastBVH | DT:         1.95326 ms | Contacts: 50
Continuous: All Convex Shape in single link
Total number of shape: 1000
   BtCastSimple | DT:         1.94631 ms | Contacts: 50
      BtCastBVH | DT:         1.73039 ms | Contacts: 50
```

**Comment by mpowelson** [comment]
@Levi-Armstrong Profile data before this change
```
Discrete: One Primitive Shape per link
Total number of shape: 1000
       BtSimple | DT:         0.06278 ms | Contacts: 50
          BtBVH | DT:         0.71566 ms | Contacts: 50
Discrete: All Primitive Shape in single link
Total number of shape: 1000
       BtSimple | DT:         0.26016 ms | Contacts: 50
          BtBVH | DT:         0.15871 ms | Contacts: 50
Discrete: One Convex Shape per link
Total number of shape: 1000
       BtSimple | DT:         1.05949 ms | Contacts: 50
          BtBVH | DT:         1.69469 ms | Contacts: 50
Discrete: All Convex Shape in single link
Total number of shape: 1000
       BtSimple | DT:         1.16004 ms | Contacts: 50
          BtBVH | DT:         1.21797 ms | Contacts: 50
Continuous: One Primitive Shape per link
Total number of shape: 1000
   BtCastSimple | DT:         2.80532 ms | Contacts: 50
      BtCastBVH | DT:         3.56515 ms | Contacts: 50
Continuous: All Primitive Shape in single link
Total number of shape: 1000
   BtCastSimple | DT:         3.25709 ms | Contacts: 50
      BtCastBVH | DT:         3.12347 ms | Contacts: 50
Continuous: One Convex Shape per link
Total number of shape: 1000
   BtCastSimple | DT:         1.19584 ms | Contacts: 50
      BtCastBVH | DT:         2.03583 ms | Contacts: 50
Continuous: All Convex Shape in single link
Total number of shape: 1000
   BtCastSimple | DT:         1.87137 ms | Contacts: 50
      BtCastBVH | DT:         1.72614 ms | Contacts: 50

```

**Comment by Levi-Armstrong** [comment]
Looks like no noticeable difference; you agree?

**Comment by mpowelson** [comment]
> Looks like no noticeable difference; you agree?

Yeah.

**Comment by Levi-Armstrong** [comment]
Could you rebase on the latest? I should be able to test this on a system tonight.

**Comment by mpowelson** [comment]
Rebased. I would expect that you should see no change.

**Review by Levi-Armstrong — APPROVED** [review]

**Comment by Levi-Armstrong** [comment]
Looks good after I test it tonight I will merge.

**Comment by mpowelson** [comment]
> Looks good after I test it tonight I will merge.

Did you get a chance to run it? Let me my if anything went wrong.

**Comment by Levi-Armstrong** [comment]
Testing it now.

**Comment by Levi-Armstrong** [comment]
No issues :+1:

**Comment by mpowelson** [comment]
Nice! I'm excited about this feature.

**Comment by Levi-Armstrong** [comment]
Me too! Thank you for working this in.

**Comment by Levi-Armstrong** [comment]
I will see if I can get this added to the SRDF.

**Pull request merged**